src/Symfony/Component/Translation/Bridge/Crowdin/CrowdinProvider.php

```diff
@@ -0,0 +1,397 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Bridge\Crowdin;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Translation\Dumper\XliffFileDumper;
+use Symfony\Component\Translation\Exception\ProviderException;
+use Symfony\Component\Translation\Loader\LoaderInterface;
+use Symfony\Component\Translation\Provider\ProviderInterface;
+use Symfony\Component\Translation\TranslatorBag;
+use Symfony\Component\Translation\TranslatorBagInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Andrii Bodnar <andrii.bodnar@crowdin.com>
+ *
+ * In Crowdin:
+ *  * Filenames refer to Symfony's translation domains;
+ *  * Identifiers refer to Symfony's translation keys;
+ *  * Translations refer to Symfony's translated messages
+ *
+ * @experimental in 5.3
+ */
+final class CrowdinProvider implements ProviderInterface
+{
+    private $client;
+    private $loader;
+    private $logger;
+    private $xliffFileDumper;
+    private $defaultLocale;
+    private $endpoint;
+    private $projectId;
+    private $filesDownloader;
+
+    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, XliffFileDumper $xliffFileDumper, string $defaultLocale, string $endpoint, int $projectId, HttpClientInterface $filesDownloader)
+    {
+        $this->client = $client;
+        $this->loader = $loader;
+        $this->logger = $logger;
+        $this->xliffFileDumper = $xliffFileDumper;
+        $this->defaultLocale = $defaultLocale;
+        $this->endpoint = $endpoint;
+        $this->projectId = $projectId;
+        $this->filesDownloader = $filesDownloader;
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('crowdin://%s', $this->endpoint);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write(TranslatorBagInterface $translatorBag): void
+    {
+        $fileList = $this->getFileList();
+
+        $responses = [];
+
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                if (0 === \count($catalogue->all($domain))) {
+                    continue;
+                }
+
+                $content = $this->xliffFileDumper->formatCatalogue($catalogue, $domain, ['default_locale' => $this->defaultLocale]);
+
+                $fileId = $this->getFileIdByDomain($fileList, $domain);
+
+                if ($catalogue->getLocale() === $this->defaultLocale) {
```

Comment:
The previous `$content` assignation is not used if `$catalogue->getLocale() !== $this->defaultLocale`. May be you don't need to loop over the catalogues of the TranslatorBag here, but as I did in other providers, find the catalogue for the `defaultLocale` or by defaut the first catalogue of the bag?

Reply:
Here I need all the locales, not only default. The default locale is used to upload/update source files in Crowdin. Other locales (not default) are used to upload translations for these source files. `$content ` is used for uploading sources and uploading translations in both cases.

Reply:
My bad, I miss the `else` statement